include/beast/websocket/detail/mask.hpp

```diff
@@ -105,73 +105,119 @@ ror(T t, unsigned n = 1)
         static_cast<typename std::make_unsigned<T>::type>(t) >> n));
 }
 
-// 32-bit Unoptimized
+// 32-bit optimized
 //
 template<class = void>
 void
-mask_inplace_general(
+mask_inplace_fast(
     boost::asio::mutable_buffer const& b,
         std::uint32_t& key)
 {
     using boost::asio::buffer_cast;
     using boost::asio::buffer_size;
-    auto const n = buffer_size(b);
+    auto n = buffer_size(b);
     auto p = buffer_cast<std::uint8_t*>(b);
+    if(n >= sizeof(key))
+    {
+        // Bring p to 4-byte alignment
+        auto const i = reinterpret_cast<
+            std::uintptr_t>(p) & (sizeof(key)-1);
+        switch(i)
+        {
+        case 1: p[2] ^= static_cast<std::uint8_t>(key >> 16);
+        case 2: p[1] ^= static_cast<std::uint8_t>(key >> 8);
+        case 3: p[0] ^= static_cast<std::uint8_t>(key);
+        {
+            auto const d = static_cast<
+                unsigned>(sizeof(key) - i);
+            key = ror(key, 8*d);
+            n -= d;
+            p += d;
+        }
+        default:
+            break;
+        }
+    }
+
+    // Mask 4 bytes at a time
     for(auto i = n / sizeof(key); i; --i)
     {
-        *p ^=  key      ; ++p;
-        *p ^= (key >> 8); ++p;
-        *p ^= (key >>16); ++p;
-        *p ^= (key >>24); ++p;
+        *reinterpret_cast<
+            std::uint32_t*>(p) ^= key;
+        p += sizeof(key);
```

Comment:
This may be faster:
```
    {
        // Mask 4 bytes at a time
        auto pp = reinterpret_cast<std::uint32_t*>(p);
        for(auto i = n / sizeof(key); i; --i)
            *pp++ ^= key;
    }
```